### ISSUE TITLE
Transforms generate same random numbers in different workers

### ISSUE THREAD

**Issue by romanziske** [TorchDSP/torchsig#343]
### Version

1.0.0 (Default)

### System Information

OS: Mac
Environment:  venv

### Description

When using transform with multiple workers, each worker generates the same sequence of random numbers. Is there a way to ensure different workers produce different random sequences?

> Initializing worker 0
> Initializing worker 1
> Worker 0 test random: 0.33351132529566385
> Worker 1 test random: 0.47123543450366023
> 
> [Worker 1] debug #1: 988.2553773375639
> [Worker 0] debug #1: 988.2553773375639
> [Worker 0] debug #2: 467.5534454036969
> [Worker 1] debug #2: 467.5534454036969
> [Worker 1] debug #3: 259.35155960134693
> [Worker 0] debug #3: 259.35155960134693

### How to Reproduce the Bug

```python
class RandomPrinter(DatasetTransform):
    """Debug transform that prints random values from a distribution.

    Useful for testing if DataLoader workers are using different random seeds.

    Args:
        value_range: Range for random values (min, max)
        name: Name to identify this instance in debug outputs
    """

    def __init__(
        self,
        value_range: Union[List, Tuple] = (0, 1000),
        name: str = "debug",
        **kwargs
    ) -> None:
        super().__init__(**kwargs)
        self.value_distribution = self.get_distribution(value_range)
        self.name = name

        # Get worker info for debugging
        worker_info = torch.utils.data.get_worker_info()
        worker_id = worker_info.id if worker_info else "main"

        # Print initialization message
        print(f"RandomPrinter '{self.name}' initialized in worker {worker_id}")

        # Set instance counter to track call sequence
        self.counter = 0

    def __call__(self, signal: DatasetSignal) -> DatasetSignal:
        """Sample from distribution and print value without modifying signal."""
        # Get random value from distribution
        value = self.value_distribution()

        # Get worker info
        worker_info = torch.utils.data.get_worker_info()
        worker_id = worker_info.id if worker_info else "main"

        # Increment counter
        self.counter += 1

        # Print debug info
        print(f"[Worker {worker_id}] {self.name} #{self.counter}: {value}")

        # Return unmodified signal
        return signal
```

**Comment by maoragai** [comment]
started to work on a solution for this, but got another error.
it seems like the solution is to specify to the dataloader the worker_init_fn function to in order to assign the seed to each worker independently.
you may look at 
[DataLoader](https://pytorch.org/docs/stable/notes/randomness.html#dataloader)
and in
[Randomness in multi-process data loading](https://pytorch.org/docs/stable/data.html#randomness-in-multi-process-data-loading)
so after modifying datamodule to pass the worker_init_fn to its dataloaders one should only specify

```python
def seed_worker(worker_id):
    worker_seed = torch.initial_seed() % 2**32* 2 + worker_id
    numpy.random.seed(worker_seed)
    random.seed(worker_seed)

narrowband_datamodule = NarrowbandDataModule(
    root = root,
    dataset_metadata = dataset_metadata,
    num_samples_train = num_samples_train,
    num_samples_val = num_samples_val,
    transforms = transforms,
    target_transforms = target_transforms,
    create_batch_size = 4,
    create_num_workers = 4,
    batch_size=4,
    num_workers=4,
    #overwrite = True
    worker_init_fn = seed_worker,
)

```

**Comment by MattCarrickPL** [comment]
Thanks @maoragai . We've got a fix that we're testing which is similar to what you've proposed. Look for a fix to be posted in 1-2 weeks.

**Comment by romanziske** [comment]
Thanks for looking into this.  I have noticed that the issue also occurs during dataset creation. When using 8 workers for example, every 8th generated wideband signal appears to be identical.

**Comment by romanziske** [comment]
note #361

**Comment by ereoh** [comment]
Fix merged with 1.1.0-dev: 625e56df9ebc7534925afcf453d05a76a2fcf023
